// backendless/es/management/data.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var Data = /*#__PURE__*/function () {
  function Data(app) {
    (0, _classCallCheck2["default"])(this, Data);
    this.app = app;
  }
  (0, _createClass2["default"])(Data, [{
    key: "createTable",
    value: function createTable(name, columns) {
      if (!name || typeof name !== 'string') {
        throw new Error('Table name must be provided and must be a string.');
      }
      if (!Array.isArray(columns)) {
        throw new Error('Columns must be a list.');
      }
      return this.app.request.post({
        url: this.app.urls.managementDataTable(),
        data: {
          name: name,
          columns: columns
        }
      });
    }
  }]);
  return Data;
}();
exports["default"] = Data;
